cirq/sim/density_matrix_simulator.py

```diff
@@ -0,0 +1,425 @@
+    (optional) way to resolve parameterized circuits, and a second which
+    takes in a list or sweep of parameter resolver:
+
+        run(circuit, param_resolver, repetitions)
+
+        run_sweep(circuit, params, repetitions)
+
+    These methods return `TrialResult`s which contain both the measurement
+    results, but also the parameters used for the parameterized
+    circuit operations. The initial state of a run is always the all 0s state
+    in the computational basis.
+
+    By contrast the simulate methods of the simulator give access to the density
+    matrix of the simulation at the end of the simulation of the circuit.
+    These methods take in two parameters that the run methods do not: a
+    qubit order and an initial state. The qubit order is necessary because an
+    ordering must be chosen for the kronecker product (see
+    `DensityMatrixTrialResult` for details of this ordering). The initial
+    state can be either the full density matrix, the full wave function (for
+    pure states), or an integer which represents the initial state of being
+    in a computational basis state for the binary representation of that
+    integer. Similar to run methods, there are two simulate methods that run
+    for single simulations or for sweeps across different parameters:
+
+        simulate(circuit, param_resolver, qubit_order, initial_state)
+
+        simulate_sweep(circuit, params, qubit_order, initial_state)
+
+    The simulate methods in contrast to the run methods do not perform
+    repetitions. The result of these simulations is a
+    `DensityMatrixTrialResult` which contains, in addition to measurement
+    results and information about the parameters that were used in the
+    simulation, access to the density matrix via the `density_matrix` method.
+
+    If one wishes to perform simulations that have access to the
+    density matrix as one steps through running the circuit there is a generator
+    which can be iterated over and each step is an object that gives access
+    to the density matrix.  This stepping through a `Circuit` is done on a
+    `Moment` by `Moment` manner.
+
+        simulate_moment_steps(circuit, param_resolver, qubit_order,
+                              initial_state)
+
+    One can iterate over the moments via
+
+        for step_result in simulate_moments(circuit):
+           # do something with the density matrix via
+           # step_result.density_matrix()
+    """
+
+    def __init__(self, dtype: Type[np.number] = np.complex64):
+        """Density matrix simulator.
+
+         Args:
+            dtype: The `numpy.dtype` used by the simulation. One of
+            `numpy.complex64` or `numpy.complex128`
+        """
+        if dtype not in {np.complex64, np.complex128}:
+            raise ValueError(
+                'dtype must be complex64 or complex128, was {}'.format(dtype))
+
+        self._dtype = dtype
+
+    def _run(self,
+        circuit: circuits.Circuit,
```

Comment:
Your editor really seems to like indenting param lists 4 spaces instead of 8 spaces. Makes them visually harder to distinguish from the statements in the body.